### PR TITLE
RavenDB-6900 For encrypted databases, replication must use SSL

### PR DESCRIPTION
**Pull request by iftahbe**
- Fix previous PR comments
- Cannot create encrypted database without using https
- Replication & ETL for encrypted databases must go through https
- Don't allow putting and distributing secret keys without https
- In raft (ConnectToPeer), if one node uses https, the other must too (commented, need to fix)

### CODE REVIEW

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 1]
```diff
@@ -635,9 +635,9 @@ private static unsafe int GetDataAndEtagTupleFromReader(TransactionOperationCont
 
         public override async Task<Stream> ConnectToPeer(string url, string apiKey)
         {
-            // TODO: This causes the tests to freeze, fix it...
-            //if (_parent.Url.StartsWith("https:", StringComparison.OrdinalIgnoreCase) != url.StartsWith("https:", StringComparison.OrdinalIgnoreCase))
-            //    throw new InvalidOperationException($"Failed to connect from node {_parent.Url} to node {url}. If HTTPS is used on one node, the other node must also use it");
+            if (_parent?.Url != null && url != null)
+                if (_parent.Url.StartsWith("https:", StringComparison.OrdinalIgnoreCase) != url.StartsWith("https:", StringComparison.OrdinalIgnoreCase))
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs]: Better to copy the value to a temp variable.

File: src/Raven.Client/Server/ETL/SqlDestination.cs [hunk 2]
```diff
@@ -51,6 +51,11 @@ public override string Name
                 return _name = $"{database}@{server} [{string.Join(" ", SqlTables.Select(x => x.TableName))}]";
             }
         }
+
+        public override bool UsingEncryptedCommunicationChannel()
+        {
+            return true; //TODO: Figure out how we can tell
```
Comment (ayende) [inline src/Raven.Client/Server/ETL/SqlDestination.cs:57]: @arek can you take a look here?

File: src/Raven.Server/Documents/Handlers/SecretKeyHandler.cs [hunk 3]
```diff
@@ -62,6 +62,9 @@ public unsafe Task PutKey()
             var name = GetStringQueryString("name");
             var overwrite = GetBoolValueQueryString("overwrite", required: false) ?? false;
 
+            if (Server.ServerCertificate == null)
```
Comment (ayende) [inline src/Raven.Server/Documents/Handlers/SecretKeyHandler.cs:65]: I'm not sure if this is correct. We might be running behind an HTTPS proxy here, so we aren't using HTTPS, but requests to us go over HTTPS.

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 4]
```diff
@@ -635,6 +635,10 @@ private static unsafe int GetDataAndEtagTupleFromReader(TransactionOperationCont
 
         public override async Task<Stream> ConnectToPeer(string url, string apiKey)
         {
+            if (_parent?.Url != null && url != null)
+                if (_parent.Url.StartsWith("https:", StringComparison.OrdinalIgnoreCase) != url.StartsWith("https:", StringComparison.OrdinalIgnoreCase))
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs:639]: Copy to a local variable, just to make sure that the Url cannot change midway

File: test/Tests.Infrastructure/RachisConsensusTestBase.cs [hunk 5]
```diff
@@ -304,13 +306,17 @@ public override bool ShouldSnapshot(Slice slice, RootObjectType type)
 
             public override async Task<Stream> ConnectToPeer(string url, string apiKey)
             {
+                if (_parent.Url.StartsWith("https:", StringComparison.OrdinalIgnoreCase) != url.StartsWith("https:", StringComparison.OrdinalIgnoreCase))
```
Comment (ayende) [inline test/Tests.Infrastructure/RachisConsensusTestBase.cs:309]: Don't you need the NRE check here as well?

File: src/Raven.Client/Server/ETL/SqlDestination.cs (hunk 2 above)
Comment (arekpalinski) [reply src/Raven.Client/Server/ETL/SqlDestination.cs:57]: I'm looking into that. BTW '@arek' isn't me :) my github login is @arekpalinski

Comment (arekpalinski) [reply src/Raven.Client/Server/ETL/SqlDestination.cs:57]: #2580

Comment (ayende) [reply src/Raven.Client/Server/ETL/SqlDestination.cs:57]: Opps, I'm using to doing that from TC

**Pull request merged**